src/main/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasher.java

```diff
@@ -0,0 +1,233 @@
+ * <p>
+ * This hasher only accepts HashFunctions that are cyclic in nature.
+ * </p>
+ * @see DynamicHasher
+ * @see ProcessType
+ */
+public class CachingHasher implements Hasher {
+
+    /**
+     * The list of byte arrays that are to be hashed.
+     */
+    private final List<long[]> buffers;
+
+    /**
+     * The hash function identity
+     */
+    private final HashFunctionIdentity functionIdentity;
+
+    /**
+     * Constructs a CachingHasher from a list of arrays of hash values.
+     * <p>
+     * The list of hash values comprises a @code{List&lt;long[]&gt;} where each @code{long[]}
+     * is comprises two (2) values that are the result of hashing the original buffer.  Thus a
+     * CachingHasher that was built from five (5) buffers will have five arrays of two @code{longs}
+     * each.
+     * </p>
+     * @param functionIdentity The identity of the function.
+     * @param buffers          a list of @code{long} arrays comprising two (2) values.
+     * @throws IllegalArgumentException if the name does not indicate a cyclic
+     *                                  hashing function.
+     */
+    public CachingHasher(HashFunctionIdentity functionIdentity, List<long[]> buffers) {
+        this.functionIdentity = checkIdentity(functionIdentity);
+        this.buffers = new ArrayList<long[]>(buffers);
+    }
+
+    /**
+     * Constructs a CachingHasher from an array of arrys of hash values.
+     *
+     * @param functionIdentity The identity of the function.
+     * @param buffers          an array of @code{long} arrays comprising two (2) values.
+     * @throws IllegalArgumentException if the name does not indicate a cyclic
+     *                                  hashing function.
+     */
+    public CachingHasher(HashFunctionIdentity functionIdentity, long[][] buffers) {
+        this.functionIdentity = checkIdentity(functionIdentity);
+        this.buffers = Arrays.asList(buffers);
+    }
+
+    /**
+     * Checks that the name is valid for this hasher.
+     *
+     * @param functionIdentity the Function Identity to check.
+     */
+    private static HashFunctionIdentity checkIdentity(HashFunctionIdentity functionIdentity) {
+        if (functionIdentity.getProcessType() != ProcessType.CYCLIC) {
+            throw new IllegalArgumentException("Only cyclic hash functions may be used in a caching hasher");
+        }
+        return functionIdentity;
+    }
+
+    @Override
+    public HashFunctionIdentity getHashFunctionIdentity() {
+        return functionIdentity;
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return buffers.isEmpty();
+    }
+
+
+    @Override
+    public PrimitiveIterator.OfInt getBits(Shape shape) {
+        HashFunctionValidator.checkAreEqual(getHashFunctionIdentity(),
+            shape.getHashFunctionIdentity());
+        return new IntIterator(shape);
+    }
+
+    /**
+     * Gets the long representations of the buffers.
+     * <p>
```

Comment:
A bit more doc would help here, i.e. the List<long[]> will contains zero or more entries, each entry is a non-null array of length two containing the 64-bit pair output from the cyclic hash function.